File: algorithms/breadth-first-traversal.ts
import { Node } from '../data-structures/binary-tree';
import { Queue } from '../data-structures/queue';

export function breadthFirstTraversal(root: Node): void {
	const queue = new Queue<Node>();
	queue.enqueue(root);

	while (!queue.isEmpty()) {
		const curr = queue.dequeue();
		console.log(curr?.val);

		if (curr?.left) {
			queue.enqueue(curr.left);
		}

		if (curr?.right) {
			queue.enqueue(curr.right);
		}
	}
}

const root = new Node(
	1,
	new Node(2, new Node(4), new Node(5)),
	new Node(3, null, new Node(6))
);

breadthFirstTraversal(root);

File: algorithms/breadth-first-traversal.go
package algorithms

import (
	"fmt"

	data_structures "github.com/claudealdric/dsa/data-structures"
)

func BfsQueue(root *data_structures.TreeNode[int]) {
	q := data_structures.NewQueue[*data_structures.TreeNode[int]]()
	if root != nil {
		q.Enqueue(root)
	}
	for q.Size() > 0 {
		curr, _ := q.Dequeue()
		fmt.Println(curr.Val)

		if curr.Left != nil {
			q.Enqueue(curr.Left)
		}

		if curr.Right != nil {
			q.Enqueue(curr.Right)
		}
	}
}
